### PR TITLE
feat: scheduler (18/): add more scheduler logic

### PR DESCRIPTION
**Pull request by michaelawyu**
### Description of your changes

This PR is part of the PRs that implement the Fleet workload scheduling.

It features logic for scheduling CRPs with a fixed set of target clusters; it also updates the logic of status handling in the scheduler framework.

I have:

- [x] Run `make reviewable` to ensure this PR is ready for review.

### How has this code been tested

- [x] Unit tests

### Special notes for your reviewer

- This PR overlaps with some of the other PRs, specifically the cluster eligibility checking logic
- Some the unit tests are kept out to control the PR size; they will be sent in a separate PR.

### CODE REVIEW

File: pkg/scheduler/framework/framework.go [hunk 1]
```diff
@@ -1126,3 +1152,180 @@ func (f *framework) runScorePlugins(ctx context.Context, state *CycleState, poli
...
+	currentCondition := meta.FindStatusCondition(policy.Status.Conditions, string(fleetv1beta1.PolicySnapshotScheduled))
+	if equalDecisions(currentDecisions, newDecisions) && condition.EqualCondition(currentCondition, &newCondition) {
+		// Skip if there is no change in decisions and conditions.
+		return nil
+	}
+
+	// Retrieve the corresponding CRP generation.
+	observedCRPGeneration, err := utils.ExtractObservedCRPGenerationFromPolicySnapshot(policy)
+	if err != nil {
+		klog.ErrorS(err, "Failed to retrieve CRP generation from annoation")
```
Comment (zhiying-lin) [inline pkg/scheduler/framework/framework.go:1254]:
```suggestion
		klog.ErrorS(err, "Failed to retrieve CRP generation from annoation", "clusterSchedulingPolicySnapshot", policyRef)
```

File: pkg/utils/common.go [hunk 2]
```diff
@@ -289,3 +289,18 @@ func ExtractSubindexFromClusterResourceSnapshot(snapshot *fleetv1beta1.ClusterRe
 
 	return true, subindex, nil
 }
+
+func ExtractObservedCRPGenerationFromPolicySnapshot(policy *fleetv1beta1.ClusterSchedulingPolicySnapshot) (int64, error) {
+	crpGeneartionStr, ok := policy.Annotations[fleetv1beta1.CRPGenerationAnnotation]
```
Comment (zhiying-lin) [inline pkg/utils/common.go:294]:
```suggestion
	crpGenerationStr, ok := policy.Annotations[fleetv1beta1.CRPGenerationAnnotation]
```

**Review by zhiying-lin — APPROVED** [review]
minor comments, others LGTM

File: pkg/scheduler/framework/framework.go [hunk 3]
```diff
@@ -625,29 +642,40 @@ func (f *framework) patchBindings(ctx context.Context, toPatch []*bindingWithPat
...
 	currentCondition := meta.FindStatusCondition(policy.Status.Conditions, string(fleetv1beta1.PolicySnapshotScheduled))
 	if equalDecisions(currentDecisions, newDecisions) && condition.EqualCondition(currentCondition, &newCondition) {
 		// Skip if there is no change in decisions and conditions.
 		return nil
 	}
 
+	// Retrieve the corresponding CRP generation.
+	observedCRPGeneration, err := utils.ExtractObservedCRPGenerationFromPolicySnapshot(policy)
+	if err != nil {
+		klog.ErrorS(err, "Failed to retrieve CRP generation from annoation")
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply pkg/scheduler/framework/framework.go]: Done.

File: pkg/scheduler/framework/framework.go (hunk 1 above)
Comment (michaelawyu) [reply pkg/scheduler/framework/framework.go]: Done.

**Pull request merged**